libusb/descriptor.c

```diff
@@ -291,23 +302,22 @@ static int parse_interface(libusb_context *ctx,
 			ifp->extra_length = len;
 		}
 
-		if (ifp->bNumEndpoints > 0) {
+		if (stashedNumEndpoints > 0) {
 			struct libusb_endpoint_descriptor *endpoint;
-			uint8_t i;
 
-			endpoint = calloc(ifp->bNumEndpoints, sizeof(*endpoint));
+			endpoint = calloc(stashedNumEndpoints, sizeof(*endpoint));
 			if (!endpoint) {
 				r = LIBUSB_ERROR_NO_MEM;
 				goto err;
 			}
 
 			ifp->endpoint = endpoint;
-			for (i = 0; i < ifp->bNumEndpoints; i++) {
-				r = parse_endpoint(ctx, endpoint + i, buffer, size);
+			for (; ifp->bNumEndpoints < stashedNumEndpoints; ifp->bNumEndpoints++) {
```

Comment:
Making this explicit with the loop ensures that the endpoints are cleared in the case of an error.